DHFPROD-4984: Filter object/array type properties using entity model

akshaysonvane opened the pull request:
### Description
Moved the logic for filtering out array/object type properties to the data service.
Also, we now use the entity model to filter out array/object type properties instead of relying on the column names sent by the FE.
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

ryanjdew commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/entitySearch/getOpticPlan.sjs:
```diff
@@ -58,6 +80,15 @@ structuredQuery = fn.head(xdmp.unquote(structuredQuery)).root;
+viewName = replaceHyphenWithUnderscore(viewName);
+schemaName = replaceHyphenWithUnderscore(schemaName);
+
 const newOptions = fn.head(xdmp.xsltEval(stylesheet, queryOptions)).root;
```
This is a tangent to this PR, but what is the minimum role assigned to the user that should be able to use this data service? I believe we'll need a ticket to add the xslt-eval privilege to that role as part of another ticket.

akshaysonvane replied:
Good point. 
This means we need to add the xslt-eval privilege to hub-central-entity-exporter role.

akshaysonvane replied:
Added a ticket: https://project.marklogic.com/jira/browse/DHFPROD-5139

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/managers/EntitySearchManager.java:
```diff
@@ -190,33 +189,35 @@ private StructuredQueryDefinition buildQuery(QueryManager queryMgr, SearchQuery
-            }
-            else { // If a property is not a Hub property, then it is an Entity Property
-                facetDef = getEntityPropertyConstraints(facetType, data, queryBuilder);
+            switch (facetType) {
```
Something to consider as this grows - a switch statement is often a bad practice because it naturally creates a large block of code with explicit knowledge of the various choices for the value being switched on. Switch statements can usually be replaced by a more "open/closed" mechanism via an interface like this:

     public interface FacetHandler 
       StructuredQueryDefinition buildQuery(FacetData data, StructuredQueryBuilder queryBuilder)

This class then has a Map<String, FacetHandler>, where String is the constraint name. And for each selected facet, it finds the matching FacetHandler and asks it to build an SQD.

The benefit of this is that you can easily unit-test each implementation of FacetHandler in isolation. And you avoid an ungainly switch statement and have a much more extensible design - when it's time to support a new constraint, you just create a new instance of FacetHandler and add it to the map. 

It may be worth making that change now, particularly if there's not enough coverage of the code in each switch block here.

akshaysonvane commented:
Moved `EntitySearchManager` class, entity search related models and switch blocks for handling facets to their own classes under `com.marklogic.hub.central.entities.search`
Also, added unit tests for all the code that was in each switch block previously.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/entities/search/FacetHandler.java:
```diff
@@ -0,0 +1,30 @@
+import com.marklogic.hub.central.entities.search.models.DocSearchQueryInfo;
+
+public interface FacetHandler {
+    String COLLECTION_CONSTRAINT_NAME = "Collection";
```
I think it's better not to define these, because this ends up causing the interface to know about all the possible implementations, which runs counter to the purpose of an interface. 

In EntitySearchManager on the develop branch, there's only one reference to each of these constants. If that's still the case, I wouldn't even bother making them constants.

akshaysonvane replied:
That's right, my bad.

I'll add it to a Constants class specific to entity search as they are used in a couple of places.

rjrudin commented on marklogic-data-hub-central/src/main/java/com/marklogic/hub/central/entities/search/impl/CreatedOnFacetHandler.java:
```diff
@@ -0,0 +1,52 @@
+
+public class CreatedOnFacetHandler implements FacetHandler {
+
+    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd");
```
This is another benefit of having a bunch of small Handler implementations - instead of packing all this state into EntitySearchManager, it's tightly associated with the Handler it belongs to.

rjrudin commented on marklogic-data-hub-central/src/test/java/com/marklogic/hub/central/entities/search/impl/CollectionFacetHandlerTest.java:
```diff
@@ -0,0 +1,31 @@
+        StructuredQueryBuilder queryBuilder = new StructuredQueryBuilder();
+
+        StructuredQueryDefinition queryDefinition = collectionFacetHandler.buildQuery(facetData, queryBuilder);
+        Fragment fragment = new Fragment(queryDefinition.serialize(), Namespace.getNamespace("sch", "http://marklogic.com/appservices/search"));
```
Sorry for the nit-pick here, but can you use "search" instead of "sch"? "search" is widely used in ML code as the prefix for this namespace.

akshaysonvane replied:
Sure.

rjrudin commented on marklogic-data-hub-central/src/test/java/com/marklogic/hub/central/entities/search/impl/CreatedOnFacetHandlerTest.java:
```diff
@@ -0,0 +1,41 @@
+        StructuredQueryBuilder queryBuilder = new StructuredQueryBuilder();
+
+        StructuredQueryDefinition queryDefinition = createdOnFacetHandler.buildQuery(facetData, queryBuilder);
+        Fragment fragment = new Fragment(queryDefinition.serialize(), Namespace.getNamespace("sch", "http://marklogic.com/appservices/search"));
```
To avoid duplication, I recommend an AbstractFacetHandlerTest that has a "protected Fragment toFragment(StructuredQueryDefinition)" method which will construct the Fragment and initialize the needed Namespace.

akshaysonvane replied:
Good idea.
